### PR TITLE
PUBDEV-5613: AutoML (and all models) should respect max_runtime_secs

### PR DESCRIPTION
**Pull request by seb-h2o**
https://0xdata.atlassian.net/browse/PUBDEV-5613
- in AutoML:
  - check for search limits before starting any training or hyperparameter search.
  - StackedEnsemble also included in those checkings.
- in ModelBuilder:
   - ensure we don't manipulate incomplete model after interrupting CV.
   - always let the final scoring happen in CV if the main model completed.

### CODE REVIEW

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 1]
```diff
@@ -487,13 +504,15 @@ private int nextInstanceCounter(String algoName, NonBlockingHashMap<String, Inte
    * Helper for hex.ModelBuilder.
    * @return
    */
-  public Job trainModel(Key<Model> key, String algoURLName, Model.Parameters parms) {
-    String algoName = ModelBuilder.algoName(algoURLName);
+  public Job<Model> trainModel(Key<Model> key, Algo algo, Model.Parameters parms) {
+    if (exceededSearchLimits(algo, key == null ? null : key.toString(), JobType.ModelBuild)) return null;
```
Comment (seb-h2o) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: I'm considering throwing an exception instead of returning null to avoid useless computations, need to see the impact first.

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 2]
```diff
@@ -338,7 +338,8 @@ public void computeCrossValidation() {
 
       _job.setReadyForView(true);
       DKV.put(_job);
-
+    } catch (Exception e) {
+      DKV.remove(_job._result); //ensure there's no incomplete model left for manipulation after crash or cancellation
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: This will only remove the object from DKV but will not call remove() on the object - is that intended

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 3]
```diff
@@ -504,29 +509,31 @@ public static void bulkBuildModels(String modelType, Job job, ModelBuilder<?, ?,
     }
     for( int i=0; i<N; ++i ) //all sub-models must be completed before the main model can be built
       try {
-        submodel_tasks[i].join();
+        final H2O.H2OCountedCompleter task = submodel_tasks[i];
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: How can the task be `null`?

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 4]
```diff
@@ -504,29 +509,31 @@ public static void bulkBuildModels(String modelType, Job job, ModelBuilder<?, ?,
...
+        if (task != null) task.join();
       } catch(RuntimeException t){
-        if(rt == null) rt = t;
+        if (rt == null) rt = t;
       }
     if(rt != null) throw rt;
   }
 
-  private void buildMainModel() {
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: Please, the changes minimal when changing core classes.

File: h2o-core/src/main/java/water/Job.java [hunk 5]
```diff
@@ -301,7 +304,6 @@ public final void update( final long newworked, final String msg) {
     @Override public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
       if(Job.isCancelledException(ex)) {
         new Barrier1OnCom().apply(Job.this);
-        _barrier = null;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/Job.java]: What is the purpose of this change?

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 6]
```diff
@@ -561,6 +561,11 @@ static public boolean contains(int[] a, int d) {
     return false;
   }
 
+  public static <T> boolean contains(T[] arr, T target) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: This method is not used in this PR.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 2 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: no, should probably call _job.remove() instead. thanks

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: not anymore, because I'm now throwing an exception instead of breaking the previous loop, but anyway, I prefer to keep the safe-check as it is so easy to reintroduce this bug.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 4 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: I know but harder to review - it is really easy to introduce a bug this way

File: h2o-core/src/main/java/water/Job.java (hunk 5 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/Job.java]: this line was just pointless (_barrier is always nullified at the end of the method) -> removed

File: h2o-core/src/main/java/water/util/ArrayUtils.java (hunk 6 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/util/ArrayUtils.java]: it is, in AutoML

File: h2o-core/src/main/java/water/Job.java (hunk 5 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/Job.java]: I think it might be a way of making sure the barrier is not accessed in this method (if it was modified). It is useless for the correct code execution, I agree with that, of course.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: do we enable asserts in prod? I'm fine with asserts instead

File: h2o-core/src/main/java/water/util/ArrayUtils.java (hunk 6 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/util/ArrayUtils.java]: I see and this is because you changed the type from String to Enum. In that case I think this method is dangerous because it uses `==` instead of `equals`. I would consider this to be a bug.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: asserts are only enabled in tests (in R h2o is also started with asserts enabled)

File: h2o-core/src/main/java/water/util/ArrayUtils.java (hunk 6 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/util/ArrayUtils.java]: good point, this was created with enum in mind, but should use equals to follow same semantic as List.contains

Comment (seb-h2o) [reply h2o-core/src/main/java/water/util/ArrayUtils.java]: by the way, there's also a bug in the contains(String[]..) version...

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 2 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: by the way, also forgot to propagate the exception here

**Comment by ledell** [comment]
@seb-h2o I tried this out from R and got an NPE:

```
library(h2o)

h2o.init()

# Import a sample binary outcome train/test set into H2O
train <- h2o.importFile("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
test <- h2o.importFile("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")

# Identify predictors and response
y <- "response"
x <- setdiff(names(train), y)

# For binary classification, response should be a factor
train[,y] <- as.factor(train[,y])
test[,y] <- as.factor(test[,y])

time <- system.time(aml <- h2o.automl(x = x, y = y, project_name = "foo",
                  training_frame = train,
                  max_runtime_secs = 300))
```
Failed at "45%":
```
  |==============================================                                                         |  45%

java.lang.NullPointerException

java.lang.NullPointerException
	at ai.h2o.automl.AutoML.learn(AutoML.java:1084)
	at ai.h2o.automl.AutoML.run(AutoML.java:347)
	at ai.h2o.automl.H2OJob$1.compute2(H2OJob.java:32)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1260)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:974)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1477)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)

Error: java.lang.NullPointerException
Timing stopped at: 2.156 0.506 303
```

**Comment by ledell** [comment]
@seb-h2o Minor change: Please edit the [AutoML User Guide](https://github.com/h2oai/h2o-3/blob/master/h2o-docs/src/product/automl.rst) rst file so that the definition of `max_models` reads:

max_models: Specify the maximum number of models to build in an AutoML run, including the two Stacked Ensemble models. Defaults to NULL/None.

Thx!

**Comment by seb-h2o** [comment]
@ledell did you remove your comment regarding AutoML user guide? I can't find it here anymore.
Anyway, just wanted to say that my latest changes are compatible with current version of the user guide.
If we want to change this behaviour and include SE into the max_models definition, it's a small change on my side and would simplify code a bit. wdyt?

**Comment by ledell** [comment]
@seb-h2o I posted that comment and then deleted it because I forgot that this is only a change to `max_runtime_secs` and not `max_models`.  We could change `max_models` to include SEs here or in a separate PR, whatever you think is easier on your end.

**Comment by seb-h2o** [comment]
@ledell, do we plan to push those changes on next release? or said differently, should I bring those to rel_wright as well?

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 7]
```diff
@@ -248,6 +249,12 @@ public void compute2() {
                           computeCrossValidation();
                           tryComplete();
                         }
+                        @Override
+                        public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
+                          Log.warn("Model training job "+_job._description+" completed with exception: "+ex);
+                          if (_job._result != null) _job._result.remove(); //ensure there's no incomplete model left for manipulation after crash or cancellation
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: If we are removing an inconsistent model - and we can expect that because we got an exception - what happens when remove itself throws an exception?

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 8]
```diff
@@ -550,9 +550,12 @@ public static String arrayToString(int[] ary) {
     return result;
   }
 
-  public static boolean contains(String[] names, String name) {
-    if (null == names) return false;
-    for (String n : names) if (n.equals(name)) return true;
+  public static <T> boolean contains(T[] arr, T target) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: Thank you

**Comment by michalkurka** [comment]
Looks good to me but please give me one more round of reviews - this is a change in core, we have to be careful ;)

**Comment by ledell** [comment]
@seb-h2o Even though this is a bug fix, I think this should wait for the 3.22 release since it will change the results (sometimes in a large way). @michalkurka What do you think?

**Comment by michalkurka** [comment]
@ledell @seb-h2o I agree - I would not put this in a fix release. We are changing the API (or the contract of the API - null vs Exception) and people who embed h2o in their software will be affected.

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 9]
```diff
@@ -539,56 +568,41 @@ private void addGridKey(Key<Grid> gridKey) {
...
    * @param searchParms hyperparameter search space
    * @return the started hyperparameter search job
    */
-  public Job<Grid> hyperparameterSearch(Key<Grid> gridKey, String algoName, Model.Parameters baseParms, Map<String, Object[]> searchParms) {
-    if (ArrayUtils.contains(skipAlgosList, algoName)) {
-      userFeedback.info(Stage.ModelTraining,"AutoML: skipping algo " + algoName + " hyperparameter search");
-      return null;
-    }
+  public Job<Grid> hyperparameterSearch(Key<Grid> gridKey, Algo algo, Model.Parameters baseParms, Map<String, Object[]> searchParms) {
+    if (exceededSearchLimits(algo, JobType.HyperparamSearch)) return null;
```
Comment (navdeep-G) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Is there a reason why the `UserFeedback` is not needed?

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 10]
```diff
@@ -539,56 +568,41 @@ private void addGridKey(Key<Grid> gridKey) {
...
+      searchCriteria.set_max_runtime_secs(Math.round(timeRemainingMs() / 1000.0));
     else
-      searchCriteria.set_max_runtime_secs(Math.min(searchCriteria.max_runtime_secs(),
-              timeRemainingMs() / 1000.0));
-
-    if (searchCriteria.max_runtime_secs() <= 0.001) {
-      userFeedback.info(Stage.ModelTraining,"AutoML: out of time; skipping " + algoName + " hyperparameter search");
-      return null;
-    }
+      searchCriteria.set_max_runtime_secs(Math.min(searchCriteria.max_runtime_secs(), Math.round(timeRemainingMs() / 1000.0)));
```
Comment (navdeep-G) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Same as above. Any reason for removing `UserFeedback`?

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 11]
```diff
@@ -661,48 +676,47 @@ private void setCommonModelBuilderParams(Model.Parameters params) {
...
-      userFeedback.info(Stage.ModelTraining,"AutoML: skipping algo " + whatWeAreSkipping + " build");
+  private boolean exceededSearchLimits(Algo algo, JobType job_type) {
+    return exceededSearchLimits(algo, null, job_type, false);
+  }
+
+  private boolean exceededSearchLimits(Algo algo, String algo_desc, JobType job_type, boolean ignoreMaxModels) {
+    String fullName = algo_desc == null ? algo.toString() : algo+" ("+algo_desc+")";
+
+    if (ArrayUtils.contains(skipAlgosList, algo)) {
+      userFeedback.info(Stage.ModelTraining,"AutoML: skipping Algo "+fullName+" in "+job_type);
```
Comment (navdeep-G) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Disregard previous comments. I see now.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 7 above)
Comment (navdeep-G) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: ^ Good point.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (navdeep-G) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: Will the assert be added?

File: h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py [hunk 12]
```diff
@@ -181,6 +181,16 @@ def prostate_automl_args():
     assert amodel.params['keep_cross_validation_fold_assignment']['actual'] == False
     assert amodel._model_json["output"]["cross_validation_fold_assignment_frame_id"] == None
 
+
+    print("Check that automl gets interrupted after `max_runtime_secs`")
+    max_runtime_secs = 30
```
Comment (navdeep-G) [inline h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py]: I don't think we had issues before with 30 sec runtime. Can try to increase this (if okay for the test harness) to really see if there are any changes.

File: h2o-r/tests/testdir_algos/automl/runit_automl_args.R [hunk 13]
```diff
@@ -25,7 +25,7 @@ automl.args.test <- function() {
   print("Check arguments to H2OAutoML class")
 
   #print("Try without a y")
-  #expect_failure(h2o.automl(training_frame = train,
+  #expect_failure(h2o.automl(training_rframe = train,
   #                          max_models = max_models,
   #                          project_name = "aml0"))
```
Comment (navdeep-G) [inline h2o-r/tests/testdir_algos/automl/runit_automl_args.R]: Is the above commented code needed?

File: h2o-r/tests/testdir_algos/automl/runit_automl_args.R [hunk 14]
```diff
@@ -246,6 +246,15 @@ automl.args.test <- function() {
...
   expect_equal(base_model@model$cross_validation_fold_assignment_frame_id, NULL)
+
+
+  print("Check that automl gets interrupted after `max_runtime_secs`")
+  max_runtime_secs <- 30
+  cancel_tolerance_secs <- 5 # should be enough for most cases given job notification mechanism
+  time <- system.time(aml19 <- h2o.automl(x=x, y=y, training_frame=train,
+                                         project_name="aml_max_runtime_secs",
+                                         max_runtime_secs=max_runtime_secs))[['elapsed']]
+  expect_lte(abs(time - max_runtime_secs), cancel_tolerance_secs)
```
Comment (navdeep-G) [inline h2o-r/tests/testdir_algos/automl/runit_automl_args.R]: Same as before. Might want to increase the max runtime.

**Comment by navdeep-G** [comment]
With these changes will `max_runtime_secs` always be respected in terms of SE models? Or will an SE always get built? @ledell

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: it's been added below

File: h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py (hunk 12 above)
Comment (seb-h2o) [reply h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py]: I'm just concerned if we want to add more tests today that wait for >5mn to complete, as part of the standard build pipeline; those would make more sense only for nightly build tests (if those did exist).
Erin and me ran individual tests with various timeouts against this branch, but if we want to add them as part of the standard build, I'm fine.

File: h2o-r/tests/testdir_algos/automl/runit_automl_args.R (hunk 13 above)
Comment (seb-h2o) [reply h2o-r/tests/testdir_algos/automl/runit_automl_args.R]: probably not, I'm usually in favour of deleting unused and/or commented code, but I saw too much of it :)

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 7 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: at the time the job executes onExceptionalCompletion, the job is already considered as completed, so I believe the new exception will be propagated to the job barrier as if the job failed entirely, but I need to double-check this and will come back with definitive answer.

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: was a very good point indeed, apparently it creates a deadlock: the jobs never completes because the exception fails to be propagating to the barrier: fix coming asap

**Comment by seb-h2o** [comment]
@navdeep-G with the changes, SE keeps ignoring max_models, but respects max_runtime_secs, so it's possible for the SE to not complete or even not run at all.

**Comment by ledell** [comment]
@navdeep-G As a follow-up to what @seb-h2o said, we will fix the `max_models` count to include SEs in a follow-up PR.

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java (hunk 9 above)
Comment (seb-h2o) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: closed

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java (hunk 10 above)
Comment (seb-h2o) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: closed

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java (hunk 11 above)
Comment (seb-h2o) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: closed

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 4 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: closed

File: h2o-core/src/main/java/water/Job.java (hunk 5 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/Job.java]: closed

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: closed

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: closed

Comment (navdeep-G) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: Great!

File: h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py (hunk 12 above)
Comment (navdeep-G) [reply h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py]: I think having nightly runs of AutoML are necessary at this point. Especially if something like this can potentially occur in the future. Thoughts @ledell ?

File: h2o-r/tests/testdir_algos/automl/runit_automl_args.R (hunk 13 above)
Comment (navdeep-G) [reply h2o-r/tests/testdir_algos/automl/runit_automl_args.R]: Ah, I see. I would preferably remove if not needed. Plus its in version control if anyone ever needed it again. :)

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 15]
```diff
@@ -252,7 +252,13 @@ public void compute2() {
                         @Override
                         public boolean onExceptionalCompletion(Throwable ex, CountedCompleter caller) {
                           Log.warn("Model training job "+_job._description+" completed with exception: "+ex);
-                          if (_job._result != null) _job._result.remove(); //ensure there's no incomplete model left for manipulation after crash or cancellation
+                          if (_job._result != null) {
+                            try {
+                              _job._result.remove(); //ensure there's no incomplete model left for manipulation after crash or cancellation
+                            } catch (Exception logged) {
+                              Log.warn("Exception thrown when removing result from job "+ _job._description, logged);
+                            }
```
[the remark this answers is not in the document]
Comment (navdeep-G) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: Great! Sounds good.

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 16]
```diff
@@ -332,13 +345,12 @@ public void computeCrossValidation() {
       // scores; compute gains/lifts
       cv_mainModelScores(N, mbs, cvModelBuilders);
 
-      // Step 7: Clean up potentially created temp frames
+      // Step 8: Clean up potentially created temp frames
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: Good catch :)

**Comment by michalkurka** [comment]
@seb-h2o I think the PR looks good from core changes point of view, thank you!
